### PR TITLE
fix(Youtube Music): timestamp fix

### PR DESCRIPTION
**Pull request by ION606**
## Description 

* Fixed the issue where the start timestamp would malfunction when changing from `use time left` mode.

This issue was caused due to the [getTimestamps](https://github.com/PreMiD/Extension/blob/release/src/util/presence/devHelper.ts#L527) function, which for SOME REASON completely ignores the first parameter in favor of using `Date.now()`

possibly resolves https://github.com/PreMiD/Presences/issues/8131

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>

image

image

<summary> Proof showing the creation/modification is working as expected </summary>

</details>

### CODE REVIEW

**Comment by veryCrunchy** [comment]
Fixed in alpha/beta

**Comment by ION606** [comment]
no, this PR deals with when the timestamp is updated (i.e. on seek)

**Comment by veryCrunchy** [comment]
Please try the alpha/beta, this doesn't seem to be an actual issue

**Comment by veryCrunchy** [comment]
I think you're better off removing the useTimeLeft toggle fully.
You're introducing more bugs for elapsed right now than that you are fixing.
Having the time left is the preferred behaviour anyway.

**Comment by ION606** [comment]
I suppose, but also, check this pull and the corresponding issue https://github.com/PreMiD/Presences/pull/8324

File: websites/Y/YouTube Music/presence.ts [hunk 1]
```diff
@@ -277,8 +283,21 @@ function updateSongTimestamps(useTimeLeft: boolean) {
...
+		// leaving this here until it gets fixed in the extension, just in case
+		// mediaTimestamps = [
+		// 	Date.now() / 1000 - presence.timestampFromFormat(currTimes),
+		// 	null,
+		// ];
+		mediaTimestamps = presence.getTimestamps(
+			presence.timestampFromFormat(currTimes),
+			null
+		);
+		delete mediaTimestamps[1];
```
Comment (veryCrunchy) [inline websites/Y/YouTube Music/presence.ts:301]: All you're doing now is setting `Math.floor(Date.now() / 1000)` in a overcomplicated fashion.
The readable equivalent would be
```suggestion
		mediaTimestamps[0] = Math.floor(Date.now() / 1000);
```
However the startTimestamp is also already being set [here](https://github.com/PreMiD/Presences/blob/40d3e76322749e1bc94b20d68062ad4b85911f0b/websites/Y/YouTube%20Music/presence.ts#L170) of the code.
So you'll have to properly take the timestampFromFormat into account

**Comment by ION606** [comment]
odd, I thought I removed that part, my fix was the toggle and the additional call on seek

image

**Comment by veryCrunchy** [comment]
If you want to have the actual elapsed time of the song you'll have to calculate the songs timestamps.
The Date.now() isn't doing much now.
Also please remove and don't set the event listeners if privacy mode is enabled.

**Comment by ION606** [comment]
> If you want to have the actual elapsed time of the song you'll have to calculate the songs timestamps. The Date.now() isn't doing much now. Also please remove and don't set the event listeners if privacy mode is enabled. And move all updateSongTimestamps calls behind the privacy mode statement.

wait where do I set event listeners?

File: websites/Y/YouTube Music/presence.ts [hunk 2]
```diff
@@ -277,8 +283,10 @@ function updateSongTimestamps(useTimeLeft: boolean) {
...
-	mediaTimestamps = presence.getTimestamps(
-		useTimeLeft ? presence.timestampFromFormat(currTimes) : Date.now(),
-		useTimeLeft ? presence.timestampFromFormat(totalTimes) : null
-	);
+	if (useTimeLeft) {
+		mediaTimestamps = presence.getTimestamps(
+			presence.timestampFromFormat(currTimes),
+			presence.timestampFromFormat(totalTimes)
+		);
+	} else mediaTimestamps[0] = Math.floor(Date.now() / 1000);
```
Comment (veryCrunchy) [inline websites/Y/YouTube Music/presence.ts:291]: Start timestamp will be "reset" each time `updateSongTimestamps` is called.
I think we want to show the elapsed time from the playing song instead.

**Review by veryCrunchy — CHANGES_REQUESTED** [review]

File: websites/Y/YouTube Music/presence.ts (hunk 2 above)
Comment (ION606) [reply websites/Y/YouTube Music/presence.ts:291]: I'm pretty sure setting the timestamp to now does nothing (it's just a theory bc it works fine for me)

Comment (veryCrunchy) [reply websites/Y/YouTube Music/presence.ts:291]: Setting the timestamp to `Date.now()` will reset the elapsed time. There's no time difference between now and now.
So each time you call it, you will reset it. which happens when a user unpauses the song or seeks.
If the user is listening to a song at 3.34 / 4.10 we want to show 3m34s elapsed, equivalent to showing 36s left.

Comment (ION606) [reply websites/Y/YouTube Music/presence.ts:291]: yer I know but it....does that. I'm confused now. Can you clone it and see if it works for you or just some strange bug for me?

Comment (veryCrunchy) [reply websites/Y/YouTube Music/presence.ts:291]: Right now you are setting Date.now to the start timestamp each time `updateSongTimestamps` is called.
I don't need to run the presence locally to see that it doesn't show the actual elapsed time of the song.

Comment (ION606) [reply websites/Y/YouTube Music/presence.ts:291]: but it....it does. Like it works perfectly fine for me

Comment (veryCrunchy) [reply websites/Y/YouTube Music/presence.ts:291]: Don't "it works on my machine" me now 😭

Comment (ION606) [reply websites/Y/YouTube Music/presence.ts:291]: HDJEJDIEJDIR IT DOES

Comment (veryCrunchy) [reply websites/Y/YouTube Music/presence.ts:291]: BUT ITS NOT THE BEHAVIOR WE WANT GOOD SIR

Comment (ION606) [reply websites/Y/YouTube Music/presence.ts:291]: OKOK IN A BIT I WILL GO OVER IT FJENIDNEIDBEIDE

Comment (veryCrunchy) [reply websites/Y/YouTube Music/presence.ts:291]: So take date.now and subtract the elapsed song time from it

File: websites/Y/YouTube Music/presence.ts [hunk 3]
```diff
@@ -275,10 +281,16 @@ function updateSongTimestamps(useTimeLeft: boolean) {
...
-		useTimeLeft ? presence.timestampFromFormat(currTimes) : Date.now(),
-		useTimeLeft ? presence.timestampFromFormat(totalTimes) : null
-	);
+		[currTimes, totalTimes] = element,
+		TSNew = presence.getTimestamps(
+			presence.timestampFromFormat(currTimes),
+			presence.timestampFromFormat(totalTimes)
+		);
+	if (useTimeLeft) mediaTimestamps = TSNew;
+	else {
```
Comment (veryCrunchy) [inline websites/Y/YouTube Music/presence.ts:290]:
```suggestion
	if (useTimeLeft) mediaTimestamps = presence.getTimestamps(
			presence.timestampFromFormat(currTimes),
			presence.timestampFromFormat(totalTimes)
		);
	else {
```
What's the reason for `TSNew`? 
It seems redundant, plus redundant `timestampFromFormat` calls when !useTimeLeft

**Review by veryCrunchy — CHANGES_REQUESTED** [review]

**Comment by veryCrunchy** [comment]
Is `useTimeLeftChanged` really necessary?

**Comment by ION606** [comment]
> Is `useTimeLeftChanged` really necessary?

yes

**Comment by veryCrunchy** [comment]
Looks good then
@theusaf (I can't request reviews ;-;)

**Review by Bas950 — APPROVED** [review]

**Review by theusaf — APPROVED** [review]

**Pull request merged**